packages/synapse-react-client/src/utils/hooks/useUploadFileEntity/usePrepareFileEntityUpload.ts

```diff
@@ -0,0 +1,113 @@
+ *
+ * In the future, this sequence could be amended to check if the storage location has enough space to accommodate all the
+ * new files, and return an error if not.
+ *
+ * @param options
+ */
+export function usePrepareFileEntityUpload(
+  options?: Partial<
+    UseMutationOptions<
+      PrepareDirsForUploadReturn,
+      SynapseClientError,
+      { files: File[]; parentId: string }
+    >
+  >,
+) {
+  const { synapseClient } = useSynapseContext()
+  const { mutateAsync: createDirsForFileList } = useCreatePathsAndGetParentId()
+
+  return useMutation({
+    ...options,
+    mutationFn: async (args: { files: File[]; parentId: string }) => {
+      const { files, parentId } = args
+
+      // 1. Create directories for the files as needed
+      const fileAndParentIds: { file: File; parentId: string }[] = []
+      for (const file of files) {
+        try {
+          // Create the directories serially; if multiple files are uploaded, they may share new directories
+          // Creating folders in parallel could cause race conditions
+          fileAndParentIds.push(await createDirsForFileList({ file, parentId }))
+        } catch (e) {
+          throw new Error(
+            `Unable to create target folder structure for file ${file.name}${
+              Object.hasOwn(e, 'message') ? `: ${e.message}` : null
+            }`,
+            { cause: e },
+          )
+        }
+      }
+
+      // 2. Check for existing files, and prompt the user if a new version should be created
+      const getExistingFilesResults = await Promise.allSettled(
+        fileAndParentIds.map(fileAndParentId =>
+          getFileEntityIdWithSameName(
+            fileAndParentId.file.name,
+            fileAndParentId.parentId,
+            synapseClient,
+            'The file could not be uploaded.',
+          ).then(existingEntityId => ({
+            ...fileAndParentId,
+            existingEntityId,
+          })),
+        ),
+      )
+
+      const filesWithError: PromiseRejectedResult[] =
+        getExistingFilesResults.filter(promise => promise.status === 'rejected')
+
+      if (filesWithError.length > 0) {
+        throw new Error(
+          `Files could not be uploaded:\n\t${filesWithError
+            .map(promise => (promise.reason as Error).message)
+            .join('\n\t')}`,
+        )
+      }
+
+      // Split the results into new and updated files
+      const filesPreparedForUpload = getExistingFilesResults
+        // All of these promises are fulfilled -- we use this filter to narrow the type
+        .filter(promise => promise.status === 'fulfilled')
+        .map(promise => promise.value)
+
+      const newFileEntities = filesPreparedForUpload.filter(
+        f => f.existingEntityId == null,
+      )
+
+      const updatedFileEntities = filesPreparedForUpload.filter(
+        f => f.existingEntityId != null,
+      )
+
+      return { newFileEntities, updatedFileEntities }
```

Comment:
This wasn't captured in any design, but I'm slightly changing how we do this in SWC.

In SWC, each file is handled serially, so the process is something like (pseudocode)

```
for file in files:
  createDirectories(file)
  const fileEntityExists = checkIfFileEntityExists(file)
  if fileEntityExists:
    promptUserToUpdateFile()
  upload(file)
```

Where now it will be something like
```
const userPrompts = []
for file in files:
  createDirectories(file)
  const fileEntityExists = checkIfFileEntityExists(file)
  if fileEntityExists:
    userPrompts.push(file)

promptUserToUpdateFiles(userPrompts)
upload(files)
```

When we prompt the user "Do you want to create a new version of file xyz?" we can give the options "Yes", "Yes to all", "No", "Cancel uploads", so their upload won't be interrupted unless there's an error.